### PR TITLE
fix(framework) Fix `aggregate_inplace()` in `strategy.py`

### PR DESCRIPTION
**Pull request by KarhouTam**
## Issue

I found that `aggregate_inplace()` actually do its job not inplace.

### Description

https://github.com/adap/flower/blob/ad811b5a0afc8bd32fb27305a8d0063f41a09ce5/src/py/flwr/server/strategy/aggregate.py#L57-L65
`scaling_factors[0] * x`, `caling_factors[i + 1] * x` would introduce additional memory overhead to temporaily store the output. Instead, using api like `np.add(x, y, out=x)` can let `numpy` write the output to `x` directly.

### Explanation

Add `out` argument of `numpy` funcition calls. 
Explanation from `numpy`: 
> out : ndarray, None, or tuple of ndarray and None, optional
A location into which the result is stored. If provided, it must have a shape that the inputs broadcast to. If not provided or None, a freshly-allocated array is returned.

However, `out` argument requires `x` and `y` has the same `dtype`, which is the most common cases. For dealing the unusual cases, I use a wrapper and set the fallback call: `np.add(x, y)` (for example).

### Proof

I use `memory_profiler` to check the memory usage.

#### Functions
```python
from memory_profiler import profile


@profile
def flwr_aggregate_inplace(results):
    """Compute in-place weighted average."""
    # Count total examples
    num_examples_total = sum(fit_res.num_examples for (_, fit_res) in results)

    # Compute scaling factors for each result
    scaling_factors = [
        fit_res.num_examples / num_examples_total for _, fit_res in results
    ]

    # Let's do in-place aggregation
    # Get first result, then add up each other
    params = [
        scaling_factors[0] * x for x in parameters_to_ndarrays(results[0][1].parameters)
    ]
    for i, (_, fit_res) in enumerate(results[1:]):
        res = (
            scaling_factors[i + 1] * x
            for x in parameters_to_ndarrays(fit_res.parameters)
        )
        params = [reduce(np.add, layer_updates) for layer_updates in zip(params, res)]

    return params


@profile
def my_aggregate_inplace(results: List[Tuple[ClientProxy, FitRes]]) -> NDArrays:
    """Compute in-place weighted average."""
    # Count total examples
    num_examples_total = sum(fit_res.num_examples for _, fit_res in results)

    # Compute scaling factors for each result
    scaling_factors = np.array(
        [fit_res.num_examples / num_examples_total for _, fit_res in results],
        dtype=np.float32,
    )

    def _try_inplace(x: NDArray, y: NDArray, np_binary_op: np.ufunc) -> Any:
        # `(x, y, out=x)` requires x and y has the same dtype, which will do the inplace-job
        return np_binary_op(x, y, out=x) if x.dtype == y.dtype else np_binary_op(x, y)

    # Let's do in-place aggregation
    # Get first result, then add up each other

    params = [
        _try_inplace(x, scaling_factors[0], np_binary_op=np.multiply)
        for x in parameters_to_ndarrays(results[0][1].parameters)
    ]
    for i, (_, fit_res) in enumerate(results[1:], start=1):
        res = (
            _try_inplace(x, scaling_factors[i], np_binary_op=np.multiply)
            for x in parameters_to_ndarrays(fit_res.parameters)
        )
        params = [
            reduce(partial(_try_inplace, np_binary_op=np.add), layer_updates)
            for layer_updates in zip(params, res)
        ]

    return params
```

#### Input
```python
test_parameters = [np.arange(8196, dtype=np.float32) for _ in range(20)]
results = [
    (
        None,
        FitRes(
            status=None,
            parameters=Parameters(
                tensors=[ndarray_to_bytes(param) for param in test_parameters],
                tensor_type="numpy.ndarray",
            ),
            num_examples=1,
            metrics=None,
        ),
    )
    for _ in range(10)
]
```
#### Main Test Entrance
```python
LOOPS = 100
def test_flwr_aggregate_inplace():
    start = time.time()
    for _ in range(LOOPS):
        flwr_aggregate_inplace(results)
    end = time.time()
    print("flwr's aggregate_inplace cost time: ", end - start)


def test_my_aggregate_inplace():
    start = time.time()
    for _ in range(LOOPS):
        my_aggregate_inplace(results)
    end = time.time()
    print("my aggregate_inplace cost time: ", end - start)


test_flwr_aggregate_inplace()
test_my_aggregate_inplace()
```

I set `LOOPS = 1` to check memory usage; disable `@profile` and set `LOOPS = 100` to check the function efficiency.

#### Results

**Memory Usage**

```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    11    118.3 MiB    118.3 MiB           1   @profile
    12                                         def flwr_aggregate_inplace(results):
    13                                             """Compute in-place weighted average."""
    14                                             # Count total examples
    15    118.3 MiB      0.0 MiB          23       num_examples_total = sum(fit_res.num_examples for (_, fit_res) in results)
    16                                         
    17                                             # Compute scaling factors for each result
    18    118.3 MiB      0.0 MiB          24       scaling_factors = [
    19    118.3 MiB      0.0 MiB          11           fit_res.num_examples / num_examples_total for _, fit_res in results
    20                                             ]
    21                                         
    22                                             # Let's do in-place aggregation
    23                                             # Get first result, then add up each other
    24    119.8 MiB      0.0 MiB          44       params = [
    25    119.8 MiB      1.4 MiB          21           scaling_factors[0] * x for x in parameters_to_ndarrays(results[0][1].parameters)
    26                                             ]
    27    120.5 MiB      0.0 MiB          10       for i, (_, fit_res) in enumerate(results[1:]):
    28    120.5 MiB      0.0 MiB         559           res = (
    29    120.5 MiB      0.0 MiB         180               scaling_factors[i + 1] * x
    30    120.5 MiB      0.3 MiB         189               for x in parameters_to_ndarrays(fit_res.parameters)
    31                                                 )
    32    120.5 MiB      0.5 MiB         207           params = [reduce(np.add, layer_updates) for layer_updates in zip(params, res)]
    33                                         
    34    120.5 MiB      0.0 MiB           1       return params


flwr's aggregate_inplace cost time:  0.09969210624694824
Filename: /root/codes/python/flower/src/py/flwr/server/strategy/test.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    37    119.5 MiB    119.5 MiB           1   @profile
    38                                         def my_aggregate_inplace(results: List[Tuple[ClientProxy, FitRes]]) -> NDArrays:
    39                                             """Compute in-place weighted average."""
    40                                             # Count total examples
    41    119.5 MiB      0.0 MiB          23       num_examples_total = sum(fit_res.num_examples for _, fit_res in results)
    42                                         
    43                                             # Compute scaling factors for each result
    44    119.5 MiB      0.0 MiB           2       scaling_factors = np.array(
    45    119.5 MiB      0.0 MiB          13           [fit_res.num_examples / num_examples_total for _, fit_res in results],
    46    119.5 MiB      0.0 MiB           1           dtype=np.float32,
    47                                             )
    48                                         
    49    120.5 MiB      0.0 MiB         384       def _try_inplace(
    50    119.5 MiB      0.0 MiB           3           x: NDArray, y: NDArray, np_binary_op: Callable[[NDArray, NDArray], NDArray]
    51                                             ):
    52                                                 # `(x, y, out=x)` requires x and y has the same dtype, which will do the inplace-job
    53    120.5 MiB      0.0 MiB         380           return np_binary_op(x, y, out=x) if x.dtype == y.dtype else np_binary_op(x, y)
    54                                         
    55                                             # Let's do in-place aggregation
    56                                             # Get first result, then add up each other
    57                                         
    58    119.5 MiB      0.0 MiB          44       params = [
    59    119.5 MiB      0.0 MiB          20           _try_inplace(x, scaling_factors[0], np_binary_op=np.multiply)
    60    119.5 MiB      0.0 MiB          21           for x in parameters_to_ndarrays(results[0][1].parameters)
    61                                             ]
    62    120.5 MiB      0.0 MiB          10       for i, (_, fit_res) in enumerate(results[1:], start=1):
    63    120.5 MiB      0.0 MiB         559           res = (
    64    120.5 MiB      0.0 MiB         180               _try_inplace(x, scaling_factors[i], np_binary_op=np.multiply)
    65    120.5 MiB      1.0 MiB         189               for x in parameters_to_ndarrays(fit_res.parameters)
    66                                                 )
    67    120.5 MiB      0.0 MiB         396           params = [
    68    120.5 MiB      0.0 MiB         180               reduce(partial(_try_inplace, np_binary_op=np.add), layer_updates)
    69    120.5 MiB      0.0 MiB         189               for layer_updates in zip(params, res)
    70                                                 ]
    71                                         
    72    120.5 MiB      0.0 MiB           1       return params
```

**Efficiency**
```
flwr's aggregate_inplace cost time:  0.7634704113006592
my aggregate_inplace cost time:  0.7479846477508545
```

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### CODE REVIEW

**Comment by jafermarq** [comment]
Hi @KarhouTam , thanks for taking a close look at that function used by many of the strategies. What's the ration of memory footprint saved by the change you propose? Is it possible to give an analytical x%  memory savings cost?

**Comment by KarhouTam** [comment]
@jafermarq Thanks for the reviewing first.

> What's the ration of memory footprint saved by the change you propose? Is it possible to give an analytical x% memory savings cost?

Sure.

**TL;DR**: Close to 50% memory savings.

The results calculated first from above used input `results` is
```
test_parameters = [np.random.randn(512, 10) for _ in range(20)]

# 10 results, each result has 10 parameters
results = [
    (
        None,
        FitRes(
            status=None,
            parameters=Parameters(
                tensors=[ndarray_to_bytes(param) for param in test_parameters],
                tensor_type="numpy.ndarray",
            ),
            num_examples=1,
            metrics=None,
        ),
    )
    for _ in range(10)
]
```
And we can see that the Flower version uses memory `2.6M` and the proposed uses `1.3M`, so 50% maybe. For confirmation, I do the test again with different inputs (`results` used in [`fedavg_test.py`](./src/py/flwr/server/strategy/fedavg_test.py)). The Flower's uses `154.4 - 136.1 = 18.3M`; and the proposed uses `145.6 - 136.7 = 8.9M`.

So I think the saving is close to 50%.

I can use the `results` variable in `fedavg_test.py` and record these two funcitons memory usage.

### Full Example

First I wrap these functions by `memory_profiler.profile`:
```python
# =============== Flower Version ==============
@profile
def flwr_aggregate_inplace(results):
    """Compute in-place weighted average."""
    # Count total examples
    num_examples_total = sum(fit_res.num_examples for (_, fit_res) in results)

    # Compute scaling factors for each result
    scaling_factors = [
        fit_res.num_examples / num_examples_total for _, fit_res in results
    ]

    # Let's do in-place aggregation
    # Get first result, then add up each other
    params = [
        scaling_factors[0] * x for x in parameters_to_ndarrays(results[0][1].parameters)
    ]
    for i, (_, fit_res) in enumerate(results[1:]):
        res = (
            scaling_factors[i + 1] * x
            for x in parameters_to_ndarrays(fit_res.parameters)
        )
        params = [reduce(np.add, layer_updates) for layer_updates in zip(params, res)]

    return params

# =============== Proposed Version ==============
@profile
def my_aggregate_inplace(results: List[Tuple[ClientProxy, FitRes]]) -> NDArrays:
    """Compute in-place weighted average."""
    # Count total examples
    num_examples_total = sum(fit_res.num_examples for _, fit_res in results)

    # Compute scaling factors for each result
    scaling_factors = [
        fit_res.num_examples / num_examples_total for _, fit_res in results
    ]

    def _try_inplace(
        x: NDArray, y: Union[NDArray, float], np_binary_op: np.ufunc
    ) -> NDArray:
        return (  # type: ignore[no-any-return]
            np_binary_op(x, y, out=x)
            if np.can_cast(y, x.dtype, casting="same_kind")
            else np_binary_op(x, np.array(y, x.dtype), out=x)
        )

    # Let's do in-place aggregation
    # Get first result, then add up each other

    params = [
        _try_inplace(x, scaling_factors[0], np_binary_op=np.multiply)
        for x in parameters_to_ndarrays(results[0][1].parameters)
    ]

    for i, (_, fit_res) in enumerate(results[1:], start=1):
        res = (
            _try_inplace(x, scaling_factors[i], np_binary_op=np.multiply)
            for x in parameters_to_ndarrays(fit_res.parameters)
        )
        params = [
            reduce(partial(_try_inplace, np_binary_op=np.add), layer_updates)
            for layer_updates in zip(params, res)
        ]

    return params
```
And set the input argument `results` as 
```python
eights0_0 = np.random.randn(100, 64)
weights0_1 = np.random.randn(314, 628, 3)
weights1_0 = np.random.randn(100, 64)
weights1_1 = np.random.randn(314, 628, 3)

results: List[Tuple[ClientProxy, FitRes]] = [
    (
        None,
        FitRes(
            status=Status(code=Code.OK, message="Success"),
            parameters=ndarrays_to_parameters([weights0_0, weights0_1]),
            num_examples=1,
            metrics={},
        ),
    ),
    (
        None,
        FitRes(
            status=Status(code=Code.OK, message="Success"),
            parameters=ndarrays_to_parameters([weights1_0, weights1_1]),
            num_examples=5,
            metrics={},
        ),
    ),
]
```
Which is the same as one in . And the test funcitons are
```python
import gc
LOOPS = 1
def test_flwr_aggregate_inplace():
    gc.disable()
    start = time.time()
    for _ in range(LOOPS):
        a = flwr_aggregate_inplace(results)
    end = time.time()
    gc.enable()
    print("flwr's aggregate_inplace cost time: ", end - start)


def test_my_aggregate_inplace():
    gc.disable()
    start = time.time()
    for _ in range(LOOPS):
        b = my_aggregate_inplace(results)
    end = time.time()
    gc.enable()
    print("my aggregate_inplace cost time: ", end - start)
```
#### Results
```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    22    136.1 MiB    136.1 MiB           1   @profile
    23                                         def flwr_aggregate_inplace(results):
    24                                             """Compute in-place weighted average."""
    25                                             # Count total examples
    26    136.1 MiB      0.0 MiB           7       num_examples_total = sum(fit_res.num_examples for (_, fit_res) in results)
    27                                         
    28                                             # Compute scaling factors for each result
    29    136.1 MiB      0.0 MiB           8       scaling_factors = [
    30    136.1 MiB      0.0 MiB           3           fit_res.num_examples / num_examples_total for _, fit_res in results
    31                                             ]
    32                                         
    33                                             # Let's do in-place aggregation
    34                                             # Get first result, then add up each other
    35    145.4 MiB      0.0 MiB           8       params = [
    36    145.4 MiB      9.2 MiB           3           scaling_factors[0] * x for x in parameters_to_ndarrays(results[0][1].parameters)
    37                                             ]
    38    154.4 MiB      0.0 MiB           2       for i, (_, fit_res) in enumerate(results[1:]):
    39    154.4 MiB      0.0 MiB           9           res = (
    40    149.9 MiB      4.6 MiB           2               scaling_factors[i + 1] * x
    41    145.4 MiB      0.0 MiB           3               for x in parameters_to_ndarrays(fit_res.parameters)
    42                                                 )
    43    154.4 MiB      4.5 MiB           5           params = [reduce(np.add, layer_updates) for layer_updates in zip(params, res)]
    44                                         
    45    154.4 MiB      0.0 MiB           1       return params


flwr's aggregate_inplace cost time:  0.03412580490112305
Filename: /root/codes/python/flower/baselines/fedrep/test.py

Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    48    136.7 MiB    136.7 MiB           1   @profile
    49                                         def my_aggregate_inplace(results: List[Tuple[ClientProxy, FitRes]]) -> NDArrays:
    50                                             """Compute in-place weighted average."""
    51                                             # Count total examples
    52    136.7 MiB      0.0 MiB           7       num_examples_total = sum(fit_res.num_examples for _, fit_res in results)
    53                                         
    54                                             # Compute scaling factors for each result
    55    136.7 MiB      0.0 MiB           8       scaling_factors = [
    56    136.7 MiB      0.0 MiB           3           fit_res.num_examples / num_examples_total for _, fit_res in results
    57                                             ]
    58                                         
    59    145.6 MiB      0.0 MiB          11       def _try_inplace(
    60    136.7 MiB      0.0 MiB           3           x: NDArray, y: Union[NDArray, float], np_binary_op: np.ufunc
    61    136.7 MiB      0.0 MiB           1       ) -> NDArray:
    62    145.6 MiB      0.0 MiB           6           return (  # type: ignore[no-any-return]
    63    145.6 MiB      0.0 MiB           6               np_binary_op(x, y, out=x)
    64    145.6 MiB      0.0 MiB           6               if np.can_cast(y, x.dtype, casting="same_kind")
    65                                                     else np_binary_op(x, np.array(y, x.dtype), out=x)
    66                                                 )
    67                                         
    68                                             # Let's do in-place aggregation
    69                                             # Get first result, then add up each other
    70                                         
    71    141.0 MiB      0.0 MiB           8       params = [
    72    141.0 MiB      0.0 MiB           2           _try_inplace(x, scaling_factors[0], np_binary_op=np.multiply)
    73    141.0 MiB      4.3 MiB           3           for x in parameters_to_ndarrays(results[0][1].parameters)
    74                                             ]
    75                                         
    76    145.6 MiB      0.0 MiB           2       for i, (_, fit_res) in enumerate(results[1:], start=1):
    77    145.6 MiB      0.0 MiB           9           res = (
    78    145.6 MiB      0.0 MiB           2               _try_inplace(x, scaling_factors[i], np_binary_op=np.multiply)
    79    145.6 MiB      4.6 MiB           3               for x in parameters_to_ndarrays(fit_res.parameters)
    80                                                 )
    81    145.6 MiB      0.0 MiB           8           params = [
    82    145.6 MiB      0.0 MiB           2               reduce(partial(_try_inplace, np_binary_op=np.add), layer_updates)
    83    145.6 MiB      0.0 MiB           3               for layer_updates in zip(params, res)
    84                                                 ]
    85                                         
    86    145.6 MiB      0.0 MiB           1       return params


my aggregate_inplace cost time:  0.008509635925292969
```

**Comment by jafermarq** [comment]
Hi @KarhouTam, this looks super good. I spend some time on Friday looking into it closely. We aim to get it merged for the next release of flower, bear with us just some days. 🙏

**Review by panh99 — DISMISSED** [review]
lgtm!

**Comment by danieljanes** [comment]
Thanks for the PR and the helpful description @KarhouTam! Great improvement.

The merge window for Flower 1.11 is pretty much closed (apart from a few scheduled PRs). We'll merge this after the 1.11 release to have it included in Flower 1.12.

**Comment by KarhouTam** [comment]
Conflicts solved. This PR is ready to be merged in anytime. 🤗

**Comment by KarhouTam** [comment]
Hi, @panh99 @danieljanes .
Is this PR still having problems? Because I found that this PR is still not merged but no new questions or reviews have been given.

**Comment by panh99** [comment]
> Hi, @panh99 @danieljanes . Is this PR still having problems? Because I found that this PR is still not merged but no new questions or reviews have been given.

Hi @KarhouTam ! We are preparing the 1.12.0 release, and your PR will be merged soon, I believe. Thanks for the reminder!

**Review by danieljanes — APPROVED** [review]

**Comment by danieljanes** [comment]
Thanks for the PR @KarhouTam, approve! This will be included in the upcoming Flower 1.12 release (ETA: end of this week or early next week)
